README.md
# randoid
Rust nanoid implementation

This is a rust implementation of [`nanoid`s](https://github.com/ai/nanoid).

It generates unique IDs as strings that are more compact than UUIDs.

By default, it generates strings of 21 characters from an alphabet of 64 symbols
(a-z, A-Z, 0-9, "_", and "-").

## Features

This particular implementation of nanoid has the following features (many of which differ from the [`nanoid` crate](https://github.com/nikolay-govorov/nanoid)):

- no_std support
- can be used without allocating (by writing characters directly to output)
- Allows using any [`Rng`](https://docs.rs/rand/latest/rand/trait.Rng.html) implementation as a source of random data.
- Implementation is optimized for the size of the alphabet being a power of 2
- [`smartstring`](https://crates.io/crates/smartstring) support, if the `smartstring` features is enabled (as an additive feature).

## Limitations

- Requires knowing the size of the alphabet at compile time (the main reason for this is it can help the compiler optimize it better)
- Size of alphabet must be a power of 2
- Use of generics could increase compilation time

If you want a more generalized alphabet that doesn't have a size that is a power of two and/or isn't know in advance, then
[`rand::distributions::Slice`](https://docs.rs/rand/0.8.5/rand/distributions/struct.Slice.html) is probably sufficient. For example:

```rust
use rand::{Rng, distributions::Slice, thread_rng};

let alphabet = ['1', '2', '3', '4', '5', '6', '7', '9', '0', 'a', 'b', 'c'];
let id: String = thread_rng().sample_iter(&Slice::new(&alphabet).unwrap()).take(21).collect();
```

## Feature Flags

- `alloc`: Requires use of the `alloc` crate, and allows creating an id as a `String`
- `std`: Use full `std` library
- `std-rand`: Inlcude `rand/std` and `rand/std_rng` features, and add support for using `thread_rng()` as the default source of random data.
- `smartstring`: Add a function for creating an id as a `SmartString`

## Usage

### Install

```toml
[dependencies]
randoid = "0.3.0"
```

### Simple

```rust
use randoid::{randoid, Generator};

// All of the below generate a string like "9wxwPU-kQU-RDjYdxj6Eq"
let id = randoid();
let id = randoid!();
let id = Generator::default().gen();
```

### Custom length

```rust

use randoid::{randoid, Generator};

// The below generate a string like "M_P_lJcWfI"
let id = randoid!(10);
let id = Generator::with_size(10).gen();
```

### Custom alphabet

```rust
use randoid::{randoid, Generator};

let id = randoid!(21, ['a', 'b', 'c', 'd']);
let id = Generator::with_alphabet(&randoid::alphabet::HEX).gen();
```

### Custom random number generator

```rust
use randoid::{randoid, Generator};
use rand::rngs::OsRng;

let id = randoid!(21, &randoid::alphabet::DEFAULT, OsRng);
let id = Generator::with_random(OsRng).gen();
```

## About the name

"nanoid" was already taken by a similar library. I considered something like "nano-id" or "nanoid2",
but thought those were too similar. Since the IDs are generated randomly, I decided on "randoid" as
an abbreviation of "RANDOm ID".

## Acknowledgments

The original [nanoid](https://github.com/ai/nanoid) of course.

Also, <https://github.com/nikolay-govorov/nanoid>, as inspiration for this project.


CHANGELOG.md
# 0.3.0

## Changes
- `Alphabet` now requires the size of the alphabet to be a power of 2 (breaking change)
- Rename `gen_id` to just `gen` (breaking change)


src/alphabet.rs
//! Definitions for the [`Alphabet`] type and common alphabets to use.
//!
//! Inlcuding the default alphabet.

/// Type for an alphabet to use for generating ids
///
/// It has a fixed length, because that can provide the compiler
/// with more optimization opportunities, and in almost all cases
/// the alphabet used will be a constant anyway.
#[derive(Debug)]
pub struct Alphabet<const N: usize = 64>(pub(crate) [char; N]);

/// Type for alphabet with 16 possible characters
pub type HexAlphabet = Alphabet<16>;

impl Default for &'static Alphabet {
    fn default() -> Self {
        &DEFAULT
    }
}

impl Default for &'static HexAlphabet {
    fn default() -> Self {
        &HEX
    }
}

impl<const N: usize> Alphabet<N> {
    /// Create a new alphabe from a set of characters
    ///
    /// The length of the array should be at least 1 and at most `u8::MAX`.
    ///
    /// Each element of the array should be unique.
    ///
    /// # Panics
    ///
    /// Panics if:
    /// - the number of character is greater than the maximum value of a u8,
    /// since no possible random byte would be able to map to some values.
    /// - the alphabet is empty, since nothing can be generated with an empty alphabet
    /// - the number of characters isn't a power of 2, because the implementation assumes a power
    /// of 2 size.
    ///
    /// # Examples
    ///
    /// ```
    /// # use randoid::Alphabet;
    /// let alph = Alphabet::new(['1', '2', '3', '4']);
    /// ```
    ///
    /// The following would panic:
    ///
    /// ```should_panic
    /// # use randoid::Alphabet;
    /// let alph = Alphabet::new([]);
    /// ```
    ///
    /// ```should_panic
    /// # use randoid::Alphabet;
    /// let c = ['0'; (u8::MAX as usize) + 1];
    /// let alph = Alphabet::new(c);
    /// ```
    ///
    /// ```should_panic
    /// # use randoid::Alphabet;
    /// let alph = Alphabet::new(['1', '2', '3']);
    /// ```
    ///
    #[track_caller]
    pub const fn new(chars: [char; N]) -> Self {
        assert!(
            N.is_power_of_two(),
            "Alphabet must have a length that is a power of two"
        );
        assert!(N != 0, "Alphabet cannot be empty");
        assert!(
            N <= u8::max_value() as usize,
            "The alphabet cannot be longer than a `u8`"
        );
        Alphabet(chars)
    }
}

/// Default alphabet for randoid
///
/// This alphabet that is safe to use in a url and uses 64 characters.
///
/// This is the default value for an alphabet of length 64
pub const DEFAULT: Alphabet = Alphabet([
    '_', '-', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f', 'g',
    'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
    'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S',
    'T', 'U', 'V', 'W', 'X', 'Y', 'Z',
]);

/// Standard hexadecimal alphabet with lowercase letters
///
/// This is the default value for an alphabet of length 16
pub const HEX: Alphabet<16> = Alphabet([
    '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f',
]);

/// Standard hexadecimal alphabet with upercase letters
pub const HEX_UPPER: Alphabet<16> = Alphabet([
    '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F',
]);


src/randfill.rs
pub trait RandomFiller {
    /// Fill a buffer with random bytes
    fn fill_random(&mut self, buf: &mut [u8]);
}

impl RandomFiller for fn(&mut [u8]) {
    fn fill_random(&mut self, buf: &mut [u8]) {
        self(buf)
    }
}

pub struct RandFn<F: FnMut(&mut [u8])>(pub F);

impl<F: FnMut(&mut [u8])> RandomFiller for RandFn<F> {
    fn fill_random(&mut self, buf: &mut [u8]) {
        self.0(buf)
    }
}

impl<F: FnMut(&mut [u8])> From<F> for RandFn<F> {
    fn from(f: F) -> Self {
        Self(f)
    }
}

#[cfg(feature = "rand")]
mod rand_impl {
    use super::RandomFiller;

    impl<'a, R: rand::Rng> RandomFiller for &'a mut R {
        fn fill_random(&mut self, buf: &mut [u8]) {
            self.fill(buf)
        }
    }

    #[derive(Clone)]
    pub struct Rng<R: rand::Rng>(pub R);

    impl<R: rand::Rng> RandomFiller for Rng<R> {
        fn fill_random(&mut self, buf: &mut [u8]) {
            self.0.fill(buf);
        }
    }

    impl<R: rand::Rng> From<R> for Rng<R> {
        fn from(rng: R) -> Self {
            Rng(rng)
        }
    }
}

#[cfg(feature = "rand")]
pub use rand_impl::*;


src/std_rand.rs
#![cfg(feature = "std-rand")]
use super::Generator;

use super::DEFAULT_SIZE;
use crate::alphabet::{Alphabet, DEFAULT};
use rand::{rngs::ThreadRng, thread_rng};

impl<'a, const N: usize> Generator<'a, ThreadRng, N> {
    /// Create a new randoid generator using a specific alphabet
    ///
    /// And using the default size and [`rand::thread_rng()`] as the RNG.
    pub fn with_alphabet(alphabet: &'a Alphabet<N>) -> Self {
        Self::new(DEFAULT_SIZE, alphabet, thread_rng())
    }
}

impl<'a> Generator<'a, ThreadRng> {
    /// Create a new randoid generator that generates ids of a specific size
    ///
    /// But use the default alphabet and [`rand::thread_rng()`] as the RNG.
    pub fn with_size(size: usize) -> Self {
        Self {
            alphabet: &DEFAULT,
            random: thread_rng(),
            size,
        }
    }
}

impl Default for Generator<'static, rand::rngs::ThreadRng> {
    fn default() -> Self {
        Self {
            alphabet: &DEFAULT,
            random: thread_rng(),
            size: DEFAULT_SIZE,
        }
    }
}

/// Simple API for generating a nano id
///
/// This creates a `String` containing a randomly generated id using the default size (21),
/// alphabet (url safe with 64 characters), and Rng (`rand::thread_rng()`).
///
/// # Example:
///
/// ```
/// use randoid::randoid;
///
/// let id = randoid();
///
/// assert_eq!(id.len(), 21);
/// for c in id.chars() {
///     assert!(c.is_alphanumeric() || c == '-' || c == '_');
/// }
/// ```
#[inline]
pub fn randoid() -> String {
    Generator::default().gen()
}


src/alphabets.rs
/// Type for an alphabet to use for generating ids
///
/// It has a fixed length, because that can provide the compiler
/// with more optimization opportunities, and in almost all cases
/// the alphabet used will be a constant anyway.
#[derive(Debug)]
pub struct Alphabet<const N: usize = 64>(pub(crate) [char; N]);

/// Type for alphabet with 16 possible characters
pub type HexAlphabet = Alphabet<16>;

impl Default for &'static Alphabet {
    fn default() -> Self {
        &Alphabet::URL
    }
}

impl Default for &'static HexAlphabet {
    fn default() -> Self {
        &HexAlphabet::HEX
    }
}

impl<const N: usize> Alphabet<N> {
    /// Create a new alphabe from a set of characters
    ///
    /// The length of the array should be at least 1 and at most `u8::MAX`.
    ///
    /// Each element of the array should be unique.
    ///
    /// # Panics
    ///
    /// Panics if the number of character is greater than the maximum value of a u8,
    /// since no possible random byte would be able to map to some values. Similarly,
    /// an empty alphabet will panic.
    ///
    /// # Examples
    ///
    /// ```
    /// # use randoid::Alphabet;
    /// let alph = Alphabet::new(['1', '2', '3', '4']);
    /// ```
    ///
    /// ```should_panic
    /// # use randoid::Alphabet;
    /// let alph = Alphabet::new([]);
    /// ```
    ///
    /// ```should_panic
    /// # use randoid::Alphabet;
    /// let c = ['0'; (u8::MAX as usize) + 1];
    /// let alph = Alphabet::new(c);
    /// ```
    pub const fn new(chars: [char; N]) -> Self {
        assert!(N != 0, "Alphabet cannot be empty");
        assert!(
            N <= u8::max_value() as usize,
            "The alphabet cannot be longer than a `u8`"
        );
        Alphabet(chars)
    }
}

impl Alphabet {
    /// Alphabet that is safe to use in a url
    ///
    /// Uses 64 characters.
    ///
    /// This is the default value for an alphabet of length 64
    pub const URL: Alphabet = Alphabet([
        '_', '-', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f',
        'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x',
        'y', 'z', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P',
        'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z',
    ]);
}

impl HexAlphabet {
    /// Standard hexadecimal alphabet with lowercase letters
    ///
    /// This is the default value for an alphabet of length 16
    pub const HEX: Alphabet<16> = Alphabet([
        '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f',
    ]);

    /// Standard hexadecimal alphabet with upercase letters
    pub const HEX_UPPER: Alphabet<16> = Alphabet([
        '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F',
    ]);
}


src/lib.rs
#![cfg_attr(not(feature = "std"), no_std)]
#![doc = include_str!("../README.md")]
#![deny(missing_docs)]

use core::cell::RefCell;
use core::fmt::{self, Write};

#[cfg(all(feature = "alloc", not(feature = "std")))]
extern crate alloc;
#[cfg(all(feature = "alloc", not(feature = "std")))]
use alloc::string::String;

pub mod alphabet;
mod std_rand;

pub use alphabet::{Alphabet, HexAlphabet};
use rand::Rng;
#[cfg(feature = "std-rand")]
pub use std_rand::*;

/// Size of the buffer to store batched random data in.
///
/// This should be big enough to fit the step size of random data
/// for up to 40 characters in the general case, and 64 if the alphabet has a
/// size that is a power of 2.
const BUFFER_SIZE: usize = 64;

/// Default length of a generated id (21)
pub const DEFAULT_SIZE: usize = 21;

///
#[derive(Clone)]
pub struct Generator<'a, R, const N: usize = 64> {
    alphabet: &'a Alphabet<N>,
    random: R,
    size: usize,
}

impl<'a, R: Rng, const N: usize> Generator<'a, R, N> {
    /// Create a new, fully specified id generator
    ///
    /// Create a new generator that genartes ids composed of `size` characters chosen at random
    /// from `alphabet`, using `random` as a source of random data.
    ///
    /// # Examples
    ///
    /// ```
    /// use randoid::{Generator, alphabet::HEX};
    /// # use rand::SeedableRng;
    ///
    /// let rand = rand_xoshiro::Xoshiro256PlusPlus::seed_from_u64(0x04040404);
    /// let mut gen = Generator::new(8, &HEX, rand);
    /// assert_eq!(gen.gen(), "905c2761");
    /// assert_eq!(gen.gen(), "304ec655");
    /// ```
    pub fn new(size: usize, alphabet: &'a Alphabet<N>, random: R) -> Self {
        Self {
            size,
            alphabet,
            random,
        }
    }

    /// Update the size of an existing generator
    ///
    /// # Example
    ///
    /// ```
    /// # use randoid::Generator;
    ///
    /// let id = Generator::default().size(32).gen();
    /// assert_eq!(id.len(), 32);
    /// ```
    pub fn size(self, size: usize) -> Self {
        Self { size, ..self }
    }

    /// Update the alphabet of an existing generator
    ///
    /// # Example
    ///
    /// ```
    /// # use randoid::{Generator, Alphabet};
    ///
    /// let id = Generator::default().alphabet(&Alphabet::new(['a', 'b', 'c', 'd'])).gen();
    /// assert!(id.chars().all(|c| matches!(c, 'a'..='d')));
    /// ```
    pub fn alphabet<const M: usize>(self, alphabet: &Alphabet<M>) -> Generator<'_, R, M> {
        Generator {
            alphabet,
            size: self.size,
            random: self.random,
        }
    }

    /// Generate a new id, and write the result to `out`
    ///
    /// This allows you to avoid creating a new string if you would simply
    /// be adding that string to something else.
    ///
    /// # Examples
    ///
    /// ```
    /// let mut ids = String::new();
    ///
    /// let mut id_gen = randoid::Generator::default();
    /// id_gen.write_to(&mut ids);
    /// ids.push('\n');
    /// id_gen.write_to(&mut ids);
    ///
    /// assert_eq!(ids.len(), 21 * 2 + 1);
    /// ```
    ///
    /// # See Also
    /// - [`Generator::fmt`]
    /// - [`Generator::gen`]
    /// - [`Generator::gen_smartstring`]
    /// - [`Generator::fmt`]
    pub fn write_to<W: Write>(&mut self, out: &mut W) -> fmt::Result {
        if self.size == 0 {
            return Ok(());
        }
        debug_assert!(N.is_power_of_two());
        let mask: usize = N - 1;
        debug_assert!(mask.count_ones() == mask.trailing_ones());
        let mut buffer = [0u8; BUFFER_SIZE];
        let mut rem = self.size;
        while rem > 0 {
            let bytes = &mut buffer[..self.size.min(BUFFER_SIZE)];
            // This generates more bits than we actually need, but using one byte per character
            // makes the implementation a lot simpler than tracking how many bits have been used.
            self.random.fill(bytes);
            for &b in &*bytes {
                let idx = b as usize & mask;
                debug_assert!(idx < N);
                // Since the alphabet size is a power of 2, applying the
                // mask ensures that idx is a valid index into the alphabet.
                out.write_char(self.alphabet.0[idx])?;
            }
            rem -= bytes.len();
        }
        Ok(())
    }

    /// Return an object which implements [`std::fmt::Display`]
    ///
    /// This allows you to pass a new generated id to `write!()`, `format!()`, etc.
    /// without having to create an intermediate string.
    ///
    /// # Warning
    ///
    /// The returned object will generate a unique id, each time it's `Display`
    /// implementation is used. It uses interior mutability in order to avoid
    /// having to store the actual id. Similarly, the random data isn't actually
    /// generated until it is written somewhere. In general I would advise against
    /// using it except as a temporary to a formatting macro.
    ///
    /// # Examples
    ///
    /// ```
    /// use randoid::Generator;
    /// # use rand::SeedableRng;
    ///
    /// let mut generator = Generator::with_random(rand_xoshiro::Xoshiro256PlusPlus::seed_from_u64(1));
    ///
    /// println!("Your new id is: {}", generator.fmt());
    ///
    /// assert_eq!(format!("uid-{}", generator.fmt()), "uid-kkb3tf6ZyJm49m5J3xuB8");
    /// let f = generator.fmt();
    ///
    /// assert_eq!(f.to_string(), "5jO6j5xWvMx17zY3e9NbN");
    /// assert_eq!(f.to_string(), "kGAK7hvw7AdqTcsFNZGtr");
    ///
    /// ```
    pub fn fmt(&mut self) -> Fmt<'_, 'a, R, N> {
        Fmt(RefCell::new(self))
    }

    /// Generate a random id as a string
    ///
    /// # Examples
    ///
    /// ```
    /// let random_id = randoid::Generator::default().gen();
    /// ```
    #[cfg(any(feature = "std", feature = "alloc"))]
    pub fn gen(&mut self) -> String {
        let mut res = String::with_capacity(self.size);
        self.write_to(&mut res).unwrap();
        res
    }

    /// Generate a random id as a smartstring
    ///
    /// # Examples
    ///
    /// ```
    /// use randoid::Generator;
    /// use smartstring::alias::String;
    ///
    /// let random_id: String = Generator::default().gen_smartstring();
    /// ```
    #[cfg(feature = "smartstring")]
    #[cfg_attr(docsrs, doc(cfg(feature = "smartstring")))]
    pub fn gen_smartstring(&mut self) -> smartstring::alias::String {
        let mut res = smartstring::alias::String::new();
        self.write_to(&mut res).unwrap();
        res
    }
}

impl<'a, R: Rng> Generator<'a, R> {
    /// Create a new randoid generator from an Rng
    ///
    /// Using the default size and alphabet
    pub fn with_random(random: R) -> Self {
        Self {
            alphabet: &alphabet::DEFAULT,
            random,
            size: DEFAULT_SIZE,
        }
    }
}

/// See [`Generator::fmt`]
pub struct Fmt<'g, 'a: 'g, R: Rng, const N: usize>(RefCell<&'g mut Generator<'a, R, N>>);

impl<'g, 'a: 'g, R: Rng, const N: usize> fmt::Display for Fmt<'g, 'a, R, N> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.0.borrow_mut().write_to(f)
    }
}

/// Convenience macro for emulating default arguments
///
/// This macro takes zero to three arguments, and generates a random id as a string.
///
/// This simulates a function with default arguments.
///
/// The first argument is the length (in characters) of the generated id. Defaults to
/// [`DEFAULT_SIZE`].
///
/// The second argument is the alphabet to use. This macro will automatically add borrow the
/// alphabet, if an owned value is passed. Defaults to [`alphabet::DEFAULT`].
///
/// The third argument is the random number generator to use. Defaults to [`rand::thread_rng()`].
///
///
/// # Examples
///
/// ```
/// use randoid::randoid;
/// use rand::SeedableRng;
/// use rand::rngs::StdRng;
///
/// // Use all defaults
/// let id = randoid!();
/// assert_eq!(id.len(), 21);
/// assert!(id.chars().all(|c| c.is_ascii_alphanumeric() || c == '-' || c == '_'));
/// // Generate id with 32 characters, but default alphabet and rng
/// let id = randoid!(32);
/// assert_eq!(id.len(), 32);
/// assert!(id.chars().all(|c| c.is_ascii_alphanumeric() || c == '-' || c == '_'));
/// // Generate id with 32 hex characters, but default rng
/// let id = randoid!(32, &randoid::alphabet::HEX);
/// assert_eq!(id.len(), 32);
/// assert!(id.chars().all(|c| c.is_ascii_hexdigit()));
/// // Generate id with 32 hex characters, using SmallRng for the RNG
/// let id = randoid!(32, &randoid::alphabet::HEX, StdRng::from_entropy());
/// assert_eq!(id.len(), 32);
/// assert!(id.chars().all(|c| c.is_ascii_hexdigit()));
///
///
/// ```
#[cfg(feature = "std-rand")]
#[macro_export]
macro_rules! randoid {
    () => {
        $crate::randoid()
    };
    ($size:expr) => {
        $crate::Generator::with_size($size).gen()
    };
    ($size:expr, &$alphabet:expr) => {
        $crate::Generator::new($size, &$alphabet, rand::thread_rng()).gen()
    };
    ($size:expr, [$($alphabet:literal),+]) => {
        randoid!($size, &$crate::alphabet::Alphabet::new([$($alphabet),+]))
    };
    ($size:expr, &$alphabet:expr, $rand:expr) => {
        $crate::Generator::new($size, &$alphabet, $rand).gen()
    };
    ($size:expr, [$($alphabet:literal),+], $rand:expr) => {
        randoid!($size, &$crate::alphabet::Alphabet::new([$($alphabet),+]), $rand)
    };
}
